annif/backend/yake.py

```diff
@@ -0,0 +1,196 @@
+                print(line, file=indexfile)
+
+    def _load_index(self, path):
+        index = dict()
+        with open(path, 'r', encoding='utf-8') as indexfile:
+            for line in indexfile:
+                label, uris = line.strip().split('\t')
+                index[label] = uris.split()
+        return index
+
+    def _create_index(self):
+        index = defaultdict(set)
+        for concept in self.graph.subjects(RDF.type, SKOS.Concept):
+            if (concept, OWL.deprecated, rdflib.Literal(True)) in self.graph:
+                continue
+            uri = str(concept)
+            for label_type in self.label_types:
+                for label in self.graph.objects(concept, label_type):
+                    if not label.language == self.params['language']:
+                        continue
+                    label = self._normalize_label(label)
+                    index[label].add(uri)
+        index.pop('', None)  # Remove possible empty string entry
+        self._index = dict(index)
+
+    def _normalize_label(self, label):
+        label = str(label)
+        if annif.util.boolean(self.params['remove_specifiers']):
+            label = re.sub(r' \(.*\)', '', label)
+        lemmatized_label = self._lemmatize_phrase(label)
+        return self._sort_phrase(lemmatized_label)
+
+    def _lemmatize_phrase(self, phrase):
+        normalized = []
+        for word in phrase.split():
+            normalized.append(
+                self.project.analyzer.normalize_word(word).lower())
+        return ' '.join(normalized)
+
+    def _sort_phrase(self, phrase):
+        words = phrase.split()
+        return ' '.join(sorted(words))
+
+    def _suggest(self, text, params):
+        self.debug(
+            f'Suggesting subjects for text "{text[:20]}..." (len={len(text)})')
+        limit = int(params['limit'])
+
+        keyphrases = self._kw_extractor.extract_keywords(text)
+        suggestions = self._keyphrases2suggestions(keyphrases)
+
+        subject_suggestions = [SubjectSuggestion(
+                uri=uri,
+                label=None,
+                notation=None,
+                score=score)
+                for uri, score in suggestions[:limit] if score > 0.0]
+        return ListSuggestionResult.create_from_index(subject_suggestions,
+                                                      self.project.subjects)
+
+    def _keyphrases2suggestions(self, keyphrases):
+        suggestions = []
+        not_matched = []
+        for kp, score in keyphrases:
+            uris = self._keyphrase2uris(kp)
+            for uri in uris:
+                suggestions.append(
+                    (uri, self._transform_score(score)))
+            if not uris:
+                not_matched.append((kp, self._transform_score(score)))
+        # Remove duplicate uris, conflating the scores
+        suggestions = self._combine_suggestions(suggestions)
+        self.debug('Keyphrases not matched:\n' + '\t'.join(
```

Comment:
In a future version, I think these non-matched keyphrases should be propagated back to the user as well, but it could be done in a subsequent PR as it requires a lot more scaffolding.